dbrp/service.go

```diff
@@ -0,0 +1,489 @@
+		}
+		b, err = bucket.Get(encodedID)
+		if err != nil {
+			return ErrDBRPNotFound
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	dbrp := &influxdb.DBRPMappingV2{}
+	if err := json.Unmarshal(b, dbrp); err != nil {
+		return nil, ErrInternalService(err)
+	}
+	// If the given orgID is wrong, it is as if we did not found a mapping scoped to this org.
+	if dbrp.OrganizationID != orgID {
+		return nil, ErrDBRPNotFound
+	}
+	// Update the default value for this DBRP.
+	if err := s.updateDefault(ctx, dbrp); err != nil {
+		return nil, ErrInternalService(err)
+	}
+	return dbrp, nil
+}
+
+// FindMany returns a list of mappings that match filter and the total count of matching dbrp mappings.
+// TODO(affo): find a smart way to apply FindOptions to a list of items.
+func (s *Service) FindMany(ctx context.Context, filter influxdb.DBRPMappingFilterV2, opts ...influxdb.FindOptions) ([]*influxdb.DBRPMappingV2, int, error) {
+	// Memoize defaults.
+	defs := make(map[string]*influxdb.ID)
+	get := func(orgID influxdb.ID, db string) (*influxdb.ID, error) {
+		k := orgID.String() + db
+		if _, ok := defs[k]; !ok {
+			id, err := s.getDefaultID(ctx, orgID, db)
+			if err != nil {
+				if kv.IsNotFound(err) {
+					defs[k] = nil
+				} else {
+					return nil, err
+				}
+			} else {
+				defs[k] = &id
+			}
+		}
+		return defs[k], nil
+	}
+
+	var err error
+	ms := []*influxdb.DBRPMappingV2{}
+
+	addDBRP := func(_, v []byte) error {
+		m := &influxdb.DBRPMappingV2{}
+		if err := json.Unmarshal(v, m); err != nil {
+			return ErrInternalService(err)
+		}
+		// Updating the Default field must be done before filtering.
+		defID, err := get(m.OrganizationID, m.Database)
+		if err != nil {
+			return ErrInternalService(err)
+		}
+		if m.ID == *defID {
+			m.Default = true
+		} else {
+			m.Default = false
+		}
+		if filterFunc(m, filter) {
+			ms = append(ms, m)
+		}
+		return nil
+	}
+
+	// Optimized path VS non-optimized one.
+	if orgID, db := filter.OrgID, filter.Database; orgID != nil && db != nil {
+		err = s.store.View(ctx, func(tx kv.Tx) error {
+			return s.byOrgAndDatabase.Walk(ctx, tx, composeForeignKey(*orgID, *db), addDBRP)
+		})
+	} else {
+		err = s.store.View(ctx, func(tx kv.Tx) error {
+			bucket, err := tx.Bucket(bucket)
+			if err != nil {
+				return ErrInternalService(err)
+			}
+			cur, err := bucket.Cursor()
+			if err != nil {
+				return ErrInternalService(err)
+			}
+
+			for k, v := cur.First(); k != nil; k, v = cur.Next() {
+				if err := addDBRP(k, v); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+	}
+
+	return ms, len(ms), err
+}
+
+// Create creates a new mapping.
+// If another mapping with same organization ID, database, and retention policy exists, an error is returned.
+// If the mapping already contains a valid ID, that one is used for storing the mapping.
+func (s *Service) Create(ctx context.Context, dbrp *influxdb.DBRPMappingV2) error {
+	if !dbrp.ID.Valid() {
+		dbrp.ID = s.IDGen.ID()
+	}
```

Comment:
It is unclear why ID can be specified or valid for a create operation. I would expect the ID must always be generated by `Service#Create`. I do note that `Create` checks whether the ID exists, and fails the operation, but the behavior is not what I would expect.

Reply:
This is mostly for testing, to mock IDs.

Is something that we agreed with @lyondhill and was already done for the tenant service.

This does not impact the HTTP side however. The ID never gets deserialized from the request, and is always generated.